lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,286 @@
+package com.meshrabiya.lib_nearby.nearby
+
+
+import android.content.Context
+import android.util.Log
+import com.google.android.gms.nearby.Nearby
+import com.google.android.gms.nearby.connection.AdvertisingOptions
+import com.google.android.gms.nearby.connection.ConnectionInfo
+import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
+import com.google.android.gms.nearby.connection.ConnectionResolution
+import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
+import com.google.android.gms.nearby.connection.DiscoveryOptions
+import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
+import com.google.android.gms.nearby.connection.Payload
+import com.google.android.gms.nearby.connection.PayloadCallback
+import com.google.android.gms.nearby.connection.PayloadTransferUpdate
+import com.google.android.gms.nearby.connection.Strategy
+import com.ustadmobile.meshrabiya.ext.addressToByteArray
+import com.ustadmobile.meshrabiya.log.MNetLogger
+import com.ustadmobile.meshrabiya.mmcp.MmcpMessage
+import com.ustadmobile.meshrabiya.mmcp.MmcpPing
+import com.ustadmobile.meshrabiya.mmcp.MmcpPong
+import com.ustadmobile.meshrabiya.vnet.VirtualPacket
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import java.net.InetAddress
+import kotlin.random.Random
+
+class NearbyVirtualNetwork(
+    context: Context,
+    private val name: String,
+    private val serviceId: String,
+    private val virtualIpAddress: Int,
+    private val broadcastAddress: Int,
+    private val strategy: Strategy = Strategy.P2P_CLUSTER,
+    private val logger: MNetLogger
+) {
+    private val connectionsClient = Nearby.getConnectionsClient(context)
+    private val desiredOutgoingConnections = 3
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    data class EndpointInfo(
+        val endpointId: String,
+        val status: EndpointStatus,
+        val ipAddress: String?,
+        val isOutgoing: Boolean
+    )
+
+    enum class EndpointStatus {
+        DISCONNECTED, CONNECTING, CONNECTED, DISCONNECTING
+    }
+
+    enum class LogLevel {
+        VERBOSE, DEBUG, INFO, WARNING, ERROR
+    }
+
+    private val _endpointStatusFlow = MutableStateFlow<Map<String, EndpointInfo>>(emptyMap())
+    val endpointStatusFlow = _endpointStatusFlow.asStateFlow()
+
+    init {
+        observeEndpointStatusFlow()
+    }
+
+    fun start() {
+        startAdvertising()
+        startDiscovery()
+    }
+
+    fun stop() {
```

Comment:
This should be close

Reply:
Close functions should be stable even if they are called multiple times - hence you need the atomic boolean.